javalib/src/main/scala/java/time/Duration.scala

```diff
@@ -0,0 +1,232 @@
+    if (millis >= -9223372036854775000L && millis <= Long.MaxValue)
+      millis.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def toNanos(): Long = {
+    // Reference implementation throws if nanos < -9223372036000000000!
+    if (totalNanos >= -9223372036000000000L && totalNanos <= Long.MaxValue)
+      totalNanos.toLong
+    else
+      throw new ArithmeticException("Long overflow")
+  }
+
+  def compareTo(that: Duration) = totalNanos.compare(that.totalNanos)
+
+  override def toString(): String = {
+    val nanos1 = if (seconds >= 0 || nanos == 0) nanos else 1000000000 - nanos
+    val secs = if (seconds >= 0 || nanos == 0) seconds else seconds + 1
+    val mins = secs / 60
+    val secsOfMin = secs % 60
+    val hours = mins / 60
+    val minsOfHour = mins % 60
+    val hourPart = if (hours == 0) "" else hours.toString + "H"
+    val minPart = if (minsOfHour == 0) "" else minsOfHour.toString + "M"
+    val decimals = f"$nanos1%09d".reverse.dropWhile(_ == '0').reverse
+    val decimalPart = if (decimals.isEmpty) "" else "." + decimals
+    val secsPart = secsOfMin match {
+      case 0 if seconds != 0 && nanos == 0 => ""
+      case 0 if seconds < 0                => "-0" + decimalPart + "S"
+      case n                               => n.toString + decimalPart + "S"
+    }
+    "PT" + hourPart + minPart + secsPart
+  }
+
+  override def equals(that: Any): Boolean = that match {
+    case that: Duration =>
+      seconds == that.getSeconds && nanos == that.getNano
+
+    case _ => false
+  }
+
+  override def hashCode(): Int = 31 * seconds.hashCode + nanos
+}
+
+object Duration {
+  private def apply(nanos: BigInt): Duration = {
+    val (quot, rem) = nanos /% 1000000000
+    val seconds = if (rem < 0) quot - 1 else quot
+    val nanosOfSeconds = if (rem < 0) rem + 1000000000 else rem
+    if (seconds.isValidLong) new Duration(seconds.toLong, nanosOfSeconds.toInt)
+    else throw new ArithmeticException("Long overflow")
+  }
+
+  val ZERO = new Duration(0, 0)
+
+  def ofDays(days: Long): Duration = Duration(BigInt(days) * 86400000000000L)
+
+  def ofHours(hours: Long): Duration = Duration(BigInt(hours) * 3600000000000L)
+
+  def ofMinutes(minutes: Long): Duration =
+    Duration(BigInt(minutes) * 60000000000L)
+
+  def ofSeconds(seconds: Long): Duration = new Duration(seconds, 0)
+
+  def ofSeconds(seconds: Long, nanoAdjustment: Long) =
+    Duration(BigInt(seconds) * 1000000000 + nanoAdjustment)
+
+  def ofMillis(millis: Long): Duration = Duration(BigInt(millis) * 1000000)
+
+  def ofNanos(nanos: Long): Duration = Duration(BigInt(nanos))
+
+  def of(amount: Long, unit: TemporalUnit): Duration = unit match {
+    case _ if !unit.isDurationEstimated =>
+      unit.getDuration.multipliedBy(amount)
+
+    case ChronoUnit.DAYS => ofDays(amount)
+
+    case _ =>
+      throw new UnsupportedTemporalTypeException(s"Unit not supported: $unit")
```

Comment:
If you try it in the REPL, you can see that it actually throws an `UnsupportedTemporalTypeException` (which is a subclass of `DateTimeException`).